src/ncp-spinel/SpinelNCPTaskJoin.cpp

```diff
@@ -255,11 +270,23 @@ nl::wpantund::SpinelNCPTaskJoin::vprocess_event(int event, va_list args)
 
 	EH_REQUIRE_WITHIN(
 		NCP_JOIN_TIMEOUT,
-		ncp_state_is_associated(mInstance->get_ncp_state()),
+		((last_status >= SPINEL_STATUS_JOIN__BEGIN) && (last_status < SPINEL_STATUS_JOIN__END))
+		|| ncp_state_is_associated(mInstance->get_ncp_state()),
 		on_error
 	);
 
-	ret = kWPANTUNDStatus_Ok;
+	ret = last_status
+		? WPANTUND_NCPERROR_TO_STATUS(last_status)
+		: kWPANTUNDStatus_Ok;
+
+	if ( (last_status == SPINEL_STATUS_JOIN_SECURITY)
+	  || (last_status == SPINEL_STATUS_JOIN_FAILURE)
+	) {
+		mInstance->change_ncp_state(CREDENTIALS_NEEDED);
+		ret = kWPANTUNDStatus_InProgress;
```

Comment:
The wpantund join command is defined to return `InProgress` if we have switched to the `credentials-needed` state. The join failed, but by supplying the appropriate credentials it will try again.